File: web/dash0/src/lib/fail-quorum.ts
import type { FailQuorum, RegionFreshness } from "@/api/hooks";

// Multi-region quorum (spec 2026-09-25-10): a client-side mirror of the
// server's resolution (server/internal/regionquorum), used by the check form to
// say what the setting means for the regions being picked. The server has the
// last word: the check detail shows its `effectiveFailQuorum` and
// `regionalIssue`, never a value derived here.

/** The form's four choices. */
export type FailQuorumMode = "default" | "all" | "majority" | "count";

/** Region count from which the default switches from "all" to "majority". */
const MAJORITY_FROM = 3;

/** Bounds of an explicit count, as the server enforces them. */
export const FAIL_QUORUM_MIN = 1;
export const FAIL_QUORUM_MAX = 100;

/** Splits a stored value into the form's mode and count. */
export function failQuorumMode(value: FailQuorum | undefined): {
  mode: FailQuorumMode;
  count: string;
} {
  if (typeof value === "number") {
    return { mode: "count", count: String(value) };
  }

  if (value === "all" || value === "majority") {
    return { mode: value, count: "" };
  }

  return { mode: "default", count: "" };
}

/**
 * The value to send, or undefined when the count is not a whole number in
 * range (the form shows an error then, and the server would refuse it).
 */
export function failQuorumValue(mode: FailQuorumMode, count: string): FailQuorum | undefined {
  if (mode !== "count") {
    return mode;
  }

  const trimmed = count.trim();
  if (!/^\d+$/.test(trimmed)) {
    return undefined;
  }

  const parsed = Number(trimmed);
  if (parsed < FAIL_QUORUM_MIN || parsed > FAIL_QUORUM_MAX) {
    return undefined;
  }

  return parsed;
}

/** How many of `regions` must be failing for the check to be down. 0 when there is no region. */
export function resolveFailQuorum(value: FailQuorum | undefined, regions: number): number {
  if (regions <= 0) {
    return 0;
  }

  if (typeof value === "number") {
    return Math.min(Math.max(value, 1), regions);
  }

  switch (value) {
    case "all":
      return regions;
    case "majority":
      return Math.floor(regions / 2) + 1;
    default:
      return regions < MAJORITY_FROM ? regions : Math.floor(regions / 2) + 1;
  }
}

/** A region reading counts as failing: down, timeout or error. */
export function isFailingRegionStatus(status: RegionFreshness["status"]): boolean {
  return status === "down" || status === "timeout" || status === "error";
}

File: server/internal/handlers/checks/fail_quorum.go
package checks

import (
	"context"
	"fmt"
	"time"

	"github.com/fclairamb/solidping/server/internal/db/models"
	"github.com/fclairamb/solidping/server/internal/regionquorum"
)

// Multi-region quorum (spec 2026-09-25-10): the API, validation and
// config-as-code half of the failQuorum setting. The incident engine half
// lives in the incidents package; both resolve through regionquorum.

// CodeInvalidFailQuorum is a failQuorum that is none of "default", "all",
// "majority" or a whole number of regions between 1 and 100.
const CodeInvalidFailQuorum = "INVALID_FAIL_QUORUM"

// fieldFailQuorum is the JSON/validation field name of the setting.
const fieldFailQuorum = "failQuorum"

// errInvalidFailQuorum is the write paths' typed error, mapped to a 400.
var errInvalidFailQuorum = regionquorum.ErrInvalid

// appendFailQuorumFinding checks the failQuorum value (shared by create,
// validate, upsert, update and the import dry run through
// requestFieldFindings).
func appendFailQuorumFinding(findings []requestFieldFinding, values *requestFieldValues) []requestFieldFinding {
	if values.FailQuorum == nil {
		return findings
	}

	if _, err := values.FailQuorum.Setting(); err != nil {
		findings = append(findings, requestFieldFinding{
			Name: fieldFailQuorum, Code: CodeInvalidFailQuorum,
			Message: errInvalidFailQuorum.Error(), Err: errInvalidFailQuorum,
		})
	}

	return findings
}

// storedFailQuorum parses a request value into the column value: nil for the
// default.
func storedFailQuorum(value *regionquorum.Value) (*string, error) {
	setting, err := value.Setting()
	if err != nil {
		return nil, errInvalidFailQuorum
	}

	return setting.Stored(), nil
}

// applyFailQuorumUpdate writes a PATCH's failQuorum onto the update. The
// default travels as an explicit clear (NULL), otherwise a check moved off
// the default could never be moved back. A passive check has no regions and
// no quorum: the value is accepted and dropped, like its regions.
func applyFailQuorumUpdate(update *models.CheckUpdate, existing *models.Check, value *regionquorum.Value) error {
	if value == nil {
		return nil
	}

	stored, err := storedFailQuorum(value)
	if err != nil {
		return err
	}

	if existing.IsPassive() {
		return nil
	}

	update.FailQuorum = stored
	update.ClearFailQuorum = stored == nil

	return nil
}

// failQuorumResponse fills failQuorum / effectiveFailQuorum on a check
// response. Both are omitted for a passive check, which has no regions.
func failQuorumResponse(check *models.Check, response *CheckResponse) {
	if check.IsPassive() {
		return
	}

	setting, _, quorum := regionquorum.ForCheck(check)
	wire := setting.Wire()
	response.FailQuorum = &wire

	if quorum > 0 {
		response.EffectiveFailQuorum = &quorum
	}
}

// exportedFailQuorum is the document value: nil (absent) for the default.
func exportedFailQuorum(check *models.Check) *regionquorum.Value {
	if check.IsPassive() {
		return nil
	}

	setting := regionquorum.FromStored(check.FailQuorum)
	if setting.Kind == regionquorum.KindDefault {
		return nil
	}

	wire := setting.Wire()

	return &wire
}

// importedFailQuorum resolves a document's failQuorum to something always
// sent on the upsert: absent means the default, sent EXPLICITLY so a re-import
// moves a check that was taken off the default back onto it (a manifest
// describes the whole state, it does not "leave it unchanged").
func importedFailQuorum(value *regionquorum.Value) *regionquorum.Value {
	if value != nil {
		out := *value

		return &out
	}

	def := regionquorum.Value(regionquorum.KeywordDefault)

	return &def
}

// failQuorumOrDefault is the canonical text of a document value for the plan
// diff: absent reads as "default", an unparseable value as itself (validation
// reports it separately).
func failQuorumOrDefault(value *regionquorum.Value) string {
	if value == nil {
		return regionquorum.KeywordDefault
	}

	setting, err := value.Setting()
	if err != nil {
		return string(*value)
	}

	return setting.String()
}

// RegionalIssueResponse is the "regional issue" block of a check detail
// (spec 2026-09-25-10): some, but fewer than the quorum, of the check's
// current regions are failing. The check's status is `warning`, and no
// incident opens. Present only while that is true.
type RegionalIssueResponse struct {
	// FailingRegions are the current regions whose newest reading fails.
	FailingRegions []string `json:"failingRegions"`
	// FailQuorum is the effective quorum: how many regions must fail for the
	// check to be down.
	FailQuorum int `json:"failQuorum"`
	// RegionCount is the check's current region count.
	RegionCount int `json:"regionCount"`
}

// BuildRegionalIssue evaluates the stored readings exactly as the incident
// engine does and returns the regional-issue block, or nil. Shared with the
// MCP diagnose tool.
func BuildRegionalIssue(
	check *models.Check, rows []models.CheckRegionState, now time.Time,
) *RegionalIssueResponse {
	_, regionCount, quorum := regionquorum.ForCheck(check)
	if !regionquorum.UsesQuorum(quorum, regionCount) {
		return nil
	}

	eval := regionquorum.Evaluate(check.Regions, regionquorum.StatesOf(rows), quorum, now, check.StaleThreshold())
	if !eval.RegionalIssue() {
		return nil
	}

	return &RegionalIssueResponse{
		FailingRegions: eval.Failing,
		FailQuorum:     quorum,
		RegionCount:    regionCount,
	}
}

// DecorateRegionFreshness adds each region's newest reading (status and
// since when it has been on that side) from check_region_states to the
// per-region freshness list. Regions with no stored reading are left as they
// are: single-region checks keep none (their status is the check's own).
func DecorateRegionFreshness(list []RegionFreshnessResponse, rows []models.CheckRegionState) {
	byRegion := make(map[string]*models.CheckRegionState, len(rows))
	for i := range rows {
		byRegion[rows[i].Region] = &rows[i]
	}

	for i := range list {
		row, ok := byRegion[list[i].Region]
		if !ok {
			continue
		}

		status := regionStatusWire(row.Status)
		since := row.StatusSince
		list[i].Status = &status
		list[i].StatusSince = &since
	}
}

// regionStatusWire is the lowercase result status of a region reading.
func regionStatusWire(status models.ResultStatus) string {
	statusInt := int(status)

	return resultStatusString(&models.Result{Status: &statusInt})
}

// attachRegionStates reads the check's per-region readings once and uses
// them for both the freshness decoration and the regional-issue block.
func (s *Service) attachRegionStates(ctx context.Context, check *models.Check, response *CheckResponse) error {
	if check.IsPassive() {
		return nil
	}

	rows, err := s.db.ListCheckRegionStates(ctx, check.UID)
	if err != nil {
		return fmt.Errorf("failed to get region states: %w", err)
	}

	DecorateRegionFreshness(response.RegionFreshness, rows)
	response.RegionalIssue = BuildRegionalIssue(check, rows, s.now())

	return nil
}
